---
title: "Detailed_Analysis_of_Differential_Expression"
author: "Bernd Daller"
format:
  html:
    code-fold: true
    embed-resources: true
    toc: true
    toc-depth: 3
    number-sections: true
execute:
  warning: false    # Suppress warnings
  error: false      # Suppress errors
  message: false    # Suppress messages
  fig-width: 15      
  fig-height: 10     
params:
  strain: 78
  nfcore_output_Path: "nf_core_output"
editor: visual
---

## SETUP AND DATA PREPARATION

### Load Libraries and Setup Environment

```{r Load libraries, echo=FALSE}
library(here)
source(file.path(here(), "R/Libraries.R"))
source(file.path(here(), "R/DifferentialExpression_helper_functions.R"))

# Process metadata
metadata <- read.csv(
  file.path(here(), "Data/PRJNA1162077_data/biosample_metadata.csv"),
  sep = ",",
  row.names = NULL
)
metadata <- metadata[order(metadata$SRR), ]
metadata[!metadata$Genotype == "WT", "Genotype"] <- "Adapted"
metadata$Genotype_CHX <- paste(metadata$Genotype, metadata$CHX, sep = "_")
metadata <- metadata %>%
  filter(Strain == params$strain) %>%
  mutate(CHX = factor(CHX)) %>%
  mutate(across(where(is.character), as.factor))
metadata$Genotype <- relevel(metadata$Genotype, ref = "WT")
metadata$CHX <- relevel(metadata$CHX, ref = "0")


# Define relevant Paths 
# Output Path
results_path <- file.path(here(), "output", paste0("Strain_", params$strain)) 
if (!dir.exists(results_path)) {
  dir.create(results_path, recursive = TRUE)
}

# Path containing Salmon quant.sf files
quant_files <- list.files(
  params$nfcore_output_Path, 
  pattern = "quant.sf$", 
  full.names = TRUE,
  recursive = TRUE
)
quant_files <- grep(params$strain, quant_files, value = TRUE) # filter for strain
sample_names <- basename(dirname(quant_files))
names(quant_files) <- sample_names


# Path containing Bakta annotated tsv files
bakta_tsv_file <- list.files(
  file.path(here(), "Data", "Bakta_annotated_files", paste0("Strain_", params$strain)), 
  full.names = TRUE,
  pattern = ".tsv",
  recursive = TRUE
)

# Path containing OrgDbs 
org_db_path <- file.path(here(), "OrgDbs")
org_db <- grep(
  params$strain, 
  list.files(org_db_path, pattern = "org.*.db", full.names = FALSE), 
  value = TRUE
)


#Define color palette
strain_colors <- list(
  '93' = c("WT_0" = "#b3b0aa", "WT_125" = "#74726a", "Adapted_0" = "#BFFCC6", "Adapted_125" = "#228B22"),
  '73' = c("WT_0" = "#b3b0aa", "WT_125" = "#74726a", "Adapted_0" = "#FFCCE6", "Adapted_125" = "#cd3363"),
  '78' = c("WT_0" = "#b3b0aa", "WT_125" = "#74726a", "Adapted_0" = "#B3D7FF", "Adapted_125" = "#4D85BD")
)
genotype_chx_colors <- strain_colors[[as.character(params$strain)]]


#set seed
set.seed(1234)
```

### Define theme 

```{r Theme}
# Define a custom theme for ggplot2
theme_custom <- function() {
  theme(
    axis.text.x = element_text(
      angle = 45,
      hjust = 1,
      size = 7,
      lineheight = 0.9,
      face = "plain"
    ),
    axis.text.y = element_text(
      size = 7,
      lineheight = 0.9,
      face = "plain"
    ),
    legend.text = element_text(
      size = 7,
      lineheight = 0.9
    ),
    legend.title = element_text(
      size = 7,
      lineheight = 0.9,
      face = "bold"
    ),
    axis.title = element_text(
      size = 9,
      lineheight = 0.9,
      face = "plain"
    ),
    plot.title = element_text(
      size = 7,
      lineheight = 0.9,
      face = "plain"
    ),
    plot.background = element_rect(
      fill = "#FFFFFF",
      colour = NA
    ),
    legend.background = element_rect(
      fill = "#FFFFFF",
      linewidth = 0.5,
      linetype = "blank"
    )
  )
}
```

### Process Gene Information and GO Annotations

```{r Process Gene Informations and GO Annotations}
# Load Bakta TSV file, skipping the first 5 lines.
bakta_tsv <- read_tsv(bakta_tsv_file, skip = 5)

# Unnest DbXrefs: Convert the DbXrefs column into a list,
# splitting entries separated by commas. Then unnest the list to create one row per DbXref.
data_long <- bakta_tsv %>%
  mutate(DbXrefs = strsplit(as.character(DbXrefs), ",\\s*")) %>%
  unnest(DbXrefs)

# Define a function to replace the last digits in GO annotations with a given set of digits.
# This is used to standardize or fix specific GO term annotations.
replace_last_digits <- function(go, num) {
  n <- nchar(num)  # Number of digits in the replacement number
  substr(go, 1, nchar(go) - n) <- substr(go, 1, nchar(go) - n)  # Modify GO term by removing last 'n' digits
  paste0(substr(go, 1, nchar(go) - n), num)  # Append the new 'num' digits to the truncated GO term
}

# Search for the correct Pannzer output file based on strain identifier, searching recursively in the GO_Pannzer2 folder.
Pannzer <- list.files(file.path(here(), "Data", "GO_Pannzer2"), pattern = paste(params$strain), recursive = T, full.names = T)
# Narrow down to the specific Pannzer file containing "GO_strain" for GO annotations.
Pannzer <- grep("GO_strain", Pannzer, value = T)

# Read GO annotations from the Pannzer output file. 
GOannotations <- read.csv(Pannzer, sep = "\t")
# Standardize GO annotations by appending GO:0000000 as a placeholder.
GOannotations$GO <- paste("GO:0000000")
# Replace the placeholder with actual GO terms by using the function defined earlier.
GOannotations$GO <- replace_last_digits(GOannotations$GO, GOannotations$goid)

# Select only relevant columns (qpid as locus tag and GO annotation).
Pannzer_GO <- GOannotations[c("qpid", "GO")]
colnames(Pannzer_GO) <- c("Locus Tag", "DbXrefs")  # Rename columns to match those in the existing dataset.

# Identify new columns from the bakta_tsv that are not present in Pannzer_GO and add them to Pannzer_GO with NA values.
newcols <- colnames(data_long)[!colnames(data_long) %in% colnames(Pannzer_GO)]
Pannzer_GO[newcols] <- NA

# Combine the Pannzer GO data with the original Bakta data, ensuring uniqueness based on Locus Tag and DbXrefs.
# The distinct function is used to avoid duplicate rows.
data_long <- rbind(data_long, Pannzer_GO) %>%
  distinct(DbXrefs, pick(contains("Locus")), .keep_all = TRUE)

# Split DbXrefs column into separate Database and Reference columns using regular expressions.
# Database is the text before the colon, Reference is the text after.
data_long <- data_long %>%
  mutate(
    Database = sub(":.*", "", DbXrefs),
    Reference = sub(".*:", "", DbXrefs)
  ) %>%
  dplyr::select(-DbXrefs)  # Remove the original DbXrefs column after splitting.

# Group by Locus and Database, then concatenate unique references within each group.
# This condenses multiple entries per locus and database pair into a single row.
data_grouped <- data_long %>%
  group_by(across(contains("Locus")), Database) %>%
  summarise(Reference = paste(unique(Reference), collapse = "; "), .groups = "drop") %>%
  left_join(data_long[!colnames(data_long) %in% "Reference"], by = c("Locus Tag", "Database")) %>%
  distinct(Database, Reference, pick(contains("Locus")), .keep_all = TRUE)

# Reshape the data from long to wide format, creating one row per Locus Tag and one column per database (wide format).
# The original bakta_tsv is joined back to ensure all relevant columns are included in the final output.
data_wide <- data_grouped %>%
  pivot_wider(id_cols = "Locus Tag", names_from = Database, values_from = Reference) %>%
  left_join(bakta_tsv[!colnames(bakta_tsv) %in% c("DbXrefs")], by = c("Locus Tag")) %>%
  distinct()

# Modify the GO column by appending the prefix "GO:" to GO annotations, converting semicolon-delimited entries to comma-separated values.
data_wide$GO <- ifelse(is.na(data_wide$GO), NA, paste0("GO:", gsub("; ", ", GO:", data_wide$GO)))

# Final preparation of gene_infos:
# Split the COG column into separate COG and COG_category columns based on semicolons.
# Then process gene locus and gene_id fields, preferring gene names where available, falling back on locus tags or product names.
gene_infos <- data_wide%>%
  mutate(across(where(is.character), as.factor)) %>%
  separate_wider_delim(COG, delim = ";", names = c("COG", "COG_category")) %>%
  exchange_cog_values() %>%
  mutate(gene_locus = ifelse(!is.na(Gene), paste0(Gene), paste0(`Locus Tag`))) %>%
  mutate(gene_id = ifelse(!is.na(Gene), paste0(Gene), paste0(Product))) %>%
  dplyr::select(gene_id, everything())  # Final column arrangement with gene_id as the first column.

```

## STRAIN ANALYSIS FOR `r params$strain`

### Principal Component Analysis

```{r PCA Overview}
# quant_files
# # Import data using tximport
tx2gene_file<-list.files(file.path(params$nfcore_output_Path),pattern = "tx2gene.tsv",full.names = T,recursive=T)
tx2gene_file<-grep(params$strain, tx2gene_file, value = T) # filter for strain
txtogene <- readr:::read_tsv(tx2gene_file, col_names = F)
tx2gene <- txtogene %>% dplyr::select(-X1)
colnames(tx2gene) <- c("TXNAME", "GENEID")



txi <- tximport(files = quant_files, type = "salmon", txOut = T, tx2gene = tx2gene)

# Create a DESeqDataSet
dds <- DESeqDataSetFromTximport(txi,
  colData = metadata,
  design = ~ Genotype + CHX + Genotype:CHX
)

# Run the differential expression analysis
dds <- DESeq(dds)
# Obtain normalized counts
normalized_counts <- counts(dds, normalized = TRUE)

# Save the DESeq object
saveRDS(dds, file.path(results_path, "dds.rds"))

# extract vsn normalized Data
vsd <- as.data.frame(assay(vst(dds, blind = T)))

PCA<-prcomp(t(vsd))

# Plot PCA
plot_pca <- autoplot(PCA, data = metadata, colour = "Genotype_CHX", shape = "Genotype", size = 5) +theme_bw()+
  theme_custom() +
  theme(
    legend.position = "top",
    legend.direction = "horizontal",
    legend.box = "vertical",
    legend.box.spacing = unit(0.5, "lines"),
    legend.title = element_text(hjust = 0.5)
  ) +
  guides(
    colour = guide_legend(order = 1,title.position = "top", title.hjust = 0,legend.title.align=0, override.aes = list(size = 3, linetype = c(0, 0, 0, 0), shape = c(17, 17, 16, 16))),
    shape = "none"
  ) +
  scale_colour_manual(values = genotype_chx_colors) +
  geom_mark_ellipse(aes(colour = Genotype_CHX)) +
  expand_limits(x = c(-0.7, 0.7), y = c(-0.7, 0.7))

plot_pca
```

### Treatment Effect in Wild-type

```{r deseq CHX125.vs.Untreated_inWT,fig.dim=c(9,7),out.width="80%"}
res_CHXresponse_WT <- results(dds, contrast=c("CHX","125", "0"))
res_CHXresponse_WT <- lfcShrink(dds, coef = list(c("CHX_125_vs_0", "GenotypeAdapted.CHX125")), type = "ashr", res = res_CHXresponse_WT)

fdr <- process_deseq(res_CHXresponse_WT, results_path = results_path, comparison = "CHX125.vs.Untreated_inWTcells", gene_infos = gene_infos, by = c("gene", "Locus Tag"))
```

#### Volcano Plot Analysis

```{r Plot Vulcano CHX125.vs.Untreated_inWT}
De_all_CHXresponse <- list.files(file.path(results_path, "CHX125.vs.Untreated_inWTcells"), pattern = "_Differential_Expression_all.xlsx", full.names = T)
Volcano_CHXresponse <- plot_volcano(De_all_CHXresponse)
Volcano_CHXresponse
```

#### Differential Expression Table

```{r DEtable CHX125.vs.Untreated_inWT}
if (nrow(fdr) == 0) {
  print("No significant genes found")
} else {
  DT::datatable(fdr[, c("gene", "gene_id", "baseMean", "log2FoldChange", "pvalue", "padj", "COG", "COG_category", "GO", "comparison")],
    extensions = "Buttons",
    options = list(
      dom = "Bfrtip",
      scrollX = TRUE,
      buttons = c("copy", "csv", "excel", "pdf", "print"),
      pageLength = 5,
      autoWidth = TRUE
    ), rownames = F, filter = "top", selection = "single", editable = T
  ) %>%
    formatRound(columns = c("baseMean", "log2FoldChange"), digits = 2) %>% # Limit baseMean and log2FoldChange to 2 decimals
    formatSignif(columns = c("pvalue", "padj"), digits = 3) # Keep pvalue and padj in scientific notation
}
```

#### Gene Set Enrichment Analysis

```{r GSEA CHX125.vs.Untreated_inWT}
gse_CHXresponse <- go_enriched_clusterprofiler(
  input_path = De_all_CHXresponse,
  ont = "BP",
  org_db = org_db
)
```

### Treatment Effect in Adapted Strains

```{r DEseq CHX125.vs.Untreated_inAdaptedcells}
res_CHXresponse <- results(dds, list(c("CHX_125_vs_0", "GenotypeAdapted.CHX125")))
res_CHXresponse <- lfcShrink(dds, coef = list(c("CHX_125_vs_0", "GenotypeAdapted.CHX125")), type = "ashr", res = res_CHXresponse)

fdr <- process_deseq(res_CHXresponse, results_path = results_path, comparison = "CHX125.vs.Untreated_inAdaptedcells", gene_infos = gene_infos, by = c("gene", "Locus Tag"))
```

#### Volcano Plot Analysis

```{r Plot Vulcano CHX125.vs.Untreated_inAdaptedcells}
De_all_CHXresponse <- list.files(file.path(results_path, "CHX125.vs.Untreated_inAdaptedcells"), pattern = "_Differential_Expression_all.xlsx", full.names = T)
Volcano_CHXresponse <- plot_volcano(De_all_CHXresponse)
Volcano_CHXresponse
```

#### Differential Expression Table

```{r DEtable CHX125.vs.Untreated_inAdaptedcells}
if (nrow(fdr) == 0) {
  print("No significant genes found")
} else {
  DT::datatable(fdr[, c("gene", "gene_id", "baseMean", "log2FoldChange", "pvalue", "padj", "COG", "COG_category", "GO", "comparison")],
    extensions = "Buttons",
    options = list(
      dom = "Bfrtip",
      scrollX = TRUE,
      buttons = c("copy", "csv", "excel", "pdf", "print"),
      pageLength = 5,
      autoWidth = TRUE
    ), rownames = F, filter = "top", selection = "single", editable = T
  ) %>%
    formatRound(columns = c("baseMean", "log2FoldChange"), digits = 2) %>% # Limit baseMean and log2FoldChange to 2 decimals
    formatSignif(columns = c("pvalue", "padj"), digits = 3) # Keep pvalue and padj in scientific notation
}
```

#### Gene Set Enrichment Analysis

```{r GSEA CHX125.vs.Untreated_inAdaptedcells}
gse_CHXresponse <- go_enriched_clusterprofiler(
  input_path = De_all_CHXresponse,
  ont = "BP",
  org_db = org_db
)
```

### Mutation Effect Without CHX Exposure

```{r DEseq Adapted.vs.WT_inCHX0Conditions}

res_MutatationEffect_CHX0 <- lfcShrink(dds, contrast = c("Genotype", "Adapted", "WT"), type = "ashr")
fdr <- process_deseq(res_MutatationEffect_CHX0, results_path = results_path, comparison = "Adapted.vs.WT_inCHX0Conditions", gene_infos = gene_infos, by = c("gene", "Locus Tag"))
```

#### Volcano Plot Analysis

```{r Plot Vulcano Adapted.vs.WT_inCHX0Conditions}
De_all_MutatationEffect_CHX0 <- list.files(file.path(results_path, "Adapted.vs.WT_inCHX0Conditions"), pattern = "_Differential_Expression_all.xlsx", full.names = T)
Volcano_MutatationEffect_CHX0 <- plot_volcano(De_all_MutatationEffect_CHX0, remove_pattern = "_inCHX0Conditions.*")
Volcano_MutatationEffect_CHX0
```

#### Differential Expression Table

```{r DEtable Adapted.vs.WT_inCHX0Conditions}
if (nrow(fdr) == 0) {
  print("No significant genes found")
} else {
  DT::datatable(fdr[, c("gene", "gene_id", "baseMean", "log2FoldChange", "pvalue", "padj", "COG", "COG_category", "GO", "comparison")],
    extensions = "Buttons",
    options = list(
      dom = "Bfrtip",
      scrollX = TRUE,
      buttons = c("copy", "csv", "excel", "pdf", "print"),
      pageLength = 5,
      autoWidth = TRUE
    ), rownames = F, filter = "top", selection = "single", editable = T
  ) %>%
    formatRound(columns = c("baseMean", "log2FoldChange"), digits = 2) %>% # Limit baseMean and log2FoldChange to 2 decimals
    formatSignif(columns = c("pvalue", "padj"), digits = 3) # Keep pvalue and padj in scientific notation
}
```

#### Gene Set Enrichment Analysis

```{r GSEA Adapted.vs.WT_inCHX0Conditions}
gse_MutatationEffect_CHX0 <- go_enriched_clusterprofiler(
  input_path = De_all_MutatationEffect_CHX0,
  ont = "BP",
  org_db = org_db
)
```

### Mutation Effect With CHX Exposure

```{r DEseq Adapted.vs.WT_inCHX125Conditions}
res_MutatationEffect_CHX125 <- results(dds, list(c("Genotype_Adapted_vs_WT", "GenotypeAdapted.CHX125")))
res_MutatationEffect_CHX125 <- lfcShrink(dds, list(c("Genotype_Adapted_vs_WT", "GenotypeAdapted.CHX125")), res = res_MutatationEffect_CHX125, type = "ashr")
fdr <- process_deseq(res_MutatationEffect_CHX125, results_path = results_path, comparison = "Adapted.vs.WT_inCHX125Conditions", gene_infos = gene_infos, by = c("gene", "Locus Tag"))
```

#### Volcano Plot Analysis

```{r Plot Vulcano Adapted.vs.WT_inCHX125Conditions}
De_all_MutatationEffect_CHX125 <- list.files(file.path(results_path, "Adapted.vs.WT_inCHX125Conditions"), pattern = "_Differential_Expression_all.xlsx", full.names = T)
Volcano_MutatationEffect_CHX125 <- plot_volcano(De_all_MutatationEffect_CHX125, remove_pattern = "_inCHX0Conditions.*")
Volcano_MutatationEffect_CHX125
```

#### Differential Expression Table

```{r DEtable Adapted.vs.WT_inCHX125Conditions}
if (nrow(fdr) == 0) {
  print("No significant genes found")
} else {
  DT::datatable(fdr[, c("gene", "gene_id", "baseMean", "log2FoldChange", "pvalue", "padj", "COG", "COG_category", "GO", "comparison")],
    extensions = "Buttons",
    options = list(
      dom = "Bfrtip",
      scrollX = TRUE,
      buttons = c("copy", "csv", "excel", "pdf", "print"),
      pageLength = 5,
      autoWidth = TRUE
    ), rownames = F, filter = "top", selection = "single", editable = T
  ) %>%
    formatRound(columns = c("baseMean", "log2FoldChange"), digits = 2) %>% # Limit baseMean and log2FoldChange to 2 decimals
    formatSignif(columns = c("pvalue", "padj"), digits = 3) # Keep pvalue and padj in scientific notation
}
```

#### Gene Set Enrichment Analysis

```{r GSEA Adapted.vs.WT_inCHX125Conditions}
gse_MutatationEffect_CHX125 <- go_enriched_clusterprofiler(
  input_path = De_all_MutatationEffect_CHX125,
  ont = "BP",
  org_db = org_db
)
```

## PUBLICATION FIGURES

### Heat Map of Selected Genes

```{r Create Heatmap}
# Load an Excel file with mutation effect data from the first sheet
Detable <- read_xlsx(De_all_MutatationEffect_CHX0, sheet = 1)

# Filter the dataset to select rows where the sequence ID is 'contig_1' and log2FoldChange > 1
# Then, sort the results by p-value and select the top 10 rows
# Finally, select only the 'gene' and 'gene_id' columns
out <- Detable %>%
  filter(`#Sequence Id` == "contig_1" & log2FoldChange > 1) %>%
  arrange(pvalue) %>%
  head(10) %>%
  dplyr::select(gene, gene_id)

# Extract gene names from the filtered data
topGenes <- out %>% dplyr::pull(gene)

# Extract and normalize the variance-stabilized data (vst) for the top genes
vst_sig <- vsd[topGenes, ]
heat <- t(scale(t(vst_sig)))  # Standardize each row (z-score normalization)

# Modify the rownames of the heatmap matrix by collapsing 'out' data into strings
rownames(heat) <- apply(out, 1, paste, collapse = " ")

# Wrap row names to a width of 60 characters for better display in the heatmap
wrapped_rownames <- sapply(rownames(heat), function(x) paste(strwrap(x, width = 30), collapse = "\n"))

# Assign the wrapped row names back to the heatmap data
rownames(heat) <- wrapped_rownames

# Define column annotations for the heatmap using metadata, showing the 'Genotype' and 'CHX' columns
col_annot <- ComplexHeatmap::HeatmapAnnotation(df = metadata[, c("Genotype", "CHX")],annotation_name_gp=gpar(fontsize = 7), show_legend = TRUE,annotation_legend_param=list(title_gp = gpar(fontsize = 7, fontface = "plain"),labels_gp = gpar(fontsize = 7)))

# Define annotation colors for 'Treatment' column
annotation_colors <- list("Treatment" = genotype_chx_colors)

# Prepare a dataframe for annotation column and assign the name 'Treatment'
annotationcolumn <- as.data.frame(metadata[, "Genotype_CHX"])
colnames(annotationcolumn) <- "Treatment"

# Define the heatmap annotation, customize legend and appearance
col_annotation <- HeatmapAnnotation(
  df = annotationcolumn,annotation_name_gp = gpar(fontsize = 7,fontface = "bold"), col = annotation_colors, annotation_name_side = "right", show_legend = T,
  annotation_legend_param = list(
    "Treatment" = list(
      title = "Genotype_CHX",   # Legend title
      direction = "horizontal", # Legend direction
      nrow = 1,                 # Number of rows for legend
      title_gp = gpar(fontsize = 7, fontface = "bold"),
      labels_gp = gpar(fontsize = 7)
    )
  ),
  which = "col"
)
# rownames(heat) <- sub(" ","\n",rownames(heat))
# rownames(heat) <- str_wrap(rownames(heat), width = 15)
# Create the heatmap using ComplexHeatmap, specify annotation, ordering, and clustering preferences
ht <- ComplexHeatmap::Heatmap(
  heat,
  name = "Z-scores", top_annotation = col_annotation, 
  cluster_rows = F, cluster_columns = F, 
  column_order = order(metadata$Genotype_CHX),
  show_column_names = F, 
  row_dend_side = "left", 
  row_names_gp = gpar(fontsize = 7), 
  border = TRUE, 
  row_names_side = "right", 
  column_title_gp = gpar(fontsize = 7,fontface = "bold"),
  column_split = factor(c(rep("WT", 6), rep("Adapted", 6)), levels = c("WT", "Adapted")), 
  heatmap_legend_param = list(
    legend_direction = "horizontal",
    legend_width = unit(2.5, "cm"),                 # Number of rows for legend
      title_gp = gpar(fontsize = 7, fontface = "bold"),
      labels_gp = gpar(fontsize = 7)
  ), 
  row_names_max_width = max_text_width(rownames(heat), gp = gpar(fontsize = 7))
)

# Capture the heatmap expression in a grid for later plotting
heatmap <- grid::grid.grabExpr(
  ComplexHeatmap::draw(ht, heatmap_legend_side = "top", annotation_legend_side = "top", 
  align_heatmap_legend = "global_center", merge_legend = TRUE), 
  padding = unit(c(2, 100, 2, 100), "mm")
)
plot_grid(heatmap)

```

### Gene Expression Plots

```{r Create Counts Plots}
de_table <- read_xlsx(De_all_MutatationEffect_CHX0)
if (length(colnames(de_table)) < 2) {
  print("nothing happened here")
  selected_genes <- list()
} else {
  selected_genes <- sapply(topGenes[1:4], function(genetype) {
    title <- genetype 
    de_table <- de_table %>%
      arrange(desc(log2FoldChange)) %>%
      filter(`#Sequence Id` == "contig_1")
    locus <- de_table[(grepl(genetype, de_table$"gene", fixed = TRUE) | 
                     grepl(genetype, de_table[, "gene_id"], fixed = TRUE)) & 
                     de_table$log2FoldChange > 0.05, "gene"]
    title <- de_table[(grepl(genetype, de_table$"gene", fixed = TRUE) | 
                     grepl(genetype, de_table[, "gene_id"], fixed = TRUE)) & 
                     de_table$log2FoldChange > 0.05, "gene_id"]
    wrapped_title <- sapply(title, function(x) paste(strwrap(x, width = 50), collapse = "\n"))

    sapply(locus, function(x) {
      FoldChange <- format(as.numeric(de_table[de_table$"gene" == x, "log2FoldChange"]), digits = 2)
      Padj <- format(as.numeric(de_table[de_table$"gene" == x, "padj"]), digits = 2)
      Counts <- plotCounts(dds, x, intgroup = c("Genotype", "CHX"), returnData = TRUE, transform = TRUE)
      Counts <- merge(
        Counts, 
        metadata[, !colnames(metadata) %in% colnames(Counts)], 
        by.x = "row.names", 
        by.y = "SRR", 
        all.x = TRUE, 
        all.y = FALSE
      )

      p <- ggplot(
        Counts, 
        aes(
          x = CHX, 
          y = count, 
          shape = Genotype, 
          color = Genotype_CHX, 
          fill = Genotype_CHX
        )
      ) +
        stat_summary(
          fun = mean,  # Changed from fun.y to fun
          geom = "crossbar", 
          linewidth = 0.5,  # Changed from size to linewidth
          aes(group = Genotype_CHX)
        ) + 
        geom_jitter(
          color = "black",
          size = 3,
          position = position_dodge2(width = 0.3)
        ) +
        facet_wrap(~Genotype) +
        theme_bw() + 
        theme_custom() +
        scale_shape_manual(values = c(21, 24)) +
        scale_fill_manual(values = genotype_chx_colors) +
        scale_color_manual(values = genotype_chx_colors) +
        guides(
          fill = guide_legend(
            order = 1,
            title.position = "left", 
            title.hjust = 0,
            legend.title.align = 0,
            override.aes = list(
              size = 3, 
              linetype = c(0, 0, 0, 0), 
              shape = c(17, 17, 16, 16), 
              color = c(genotype_chx_colors[order(names(genotype_chx_colors))]), 
              fill = c(NA, NA, NA, NA)
            )
          ),
          shape = "none", 
          color = "none"
        ) +
        labs(
          title = wrapped_title,
          x = NULL,
          y = "counts"
        ) +
        theme(plot.title = element_text(size = 7)) + 
        expand_limits(y = c(0, ceiling(max(Counts$count) + 0.2 * max(Counts$count))))
    }, simplify = FALSE)
  }, simplify = FALSE)
}

selected_genes <- unlist(selected_genes, recursive = F)

# Extract the legend from one of the selected gene plots
legend_boxplots <- get_legend(selected_genes[[1]] + theme(legend.position = "bottom"))

# Create individual plots for selected genes without the legend
top <- lapply(seq_along(selected_genes), function(x) {
  y <- selected_genes[[x]]+ theme(legend.position = "none")
  return(y)
})

# Combine the selected gene plots into a grid with 2 columns
expression <- cowplot::plot_grid(plotlist = top, ncol = 2)+
  theme(plot.background = element_rect(fill = "#FFFFFF", colour = NA))
expression
```

### GO Enrichment and CHX Effect

```{r Create GSEA Plot}
# New facet label names for supp variable
supp.labs <- c("activated\nin\nadapted", "activated\nin\nWT")
names(supp.labs) <- c("activated", "suppressed")
gse_df<-as.data.frame(gse_MutatationEffect_CHX0)

GSE <- dotplot(gse_MutatationEffect_CHX0, showCategory = 5, split = ".sign") + coord_flip() + facet_grid(.sign ~ ., labeller = labeller(.rows = supp.labs)) +theme_custom()+ theme(legend.box = "horizontal", axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))

data<-GSE$data[order(GSE$data$GeneRatio,decreasing=T),]
data<-data[1:5,]
GSE <- GSE+
  expand_limits(x = c(0.2,1)) 


right_side <- plot_grid(heatmap, expression, legend_boxplots, labels = c("C", "D"), label_size = 10, ncol = 1, rel_heights = c(1.3, 1.3, 0.1), scale = c(0.98)) 


left_side <- plot_grid(plot_pca, Volcano_MutatationEffect_CHX0 + theme(legend.position = "none", legend.text = element_text(size = 6)) + guides(color = guide_legend(override.aes = list(size = 1.5))), labels = c("A", "B"), label_size = 10, ncol = 1, rel_heights = c(1, 1.7), scale = c(0.98)) 
final_sample <- plot_grid(left_side, right_side, rel_widths = c(1, 1.2), label_size = 10, ncol = 2) 
final_sample <- plot_grid(final_sample, GSE, ncol = 1, rel_heights = c(3, 1.3), scale = c(1, 0.98), labels = c("", "E"), label_size = 10)+ theme(plot.background = element_rect(fill = "#FFFFFF", colour = NA))

ggsave(final_sample, file = file.path(results_path, paste0("final_sample_GSE.tif")), dpi = 600, width = 10, height = 11)
final_sample
```

### Strain Specific Analysis

```{r CHX Strain 93}
if (params$strain == 93) {
  Treatment_adapted <- plot_volcano(De_all_CHXresponse)
  Treatment_adapted

  GSE <- dotplot(gse_CHXresponse, showCategory = 5, split = ".sign") + facet_grid(. ~ .sign) + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    expand_limits(x = c(0.1, 0.1))

  CHX_effect_93 <- plot_grid(
    plot_grid(Treatment_adapted + theme(legend.position = "none"), GSE, labels = c("A", "B"), label_size = 10, scale = 0.98))+   
    theme(plot.background = element_rect(fill = "#FFFFFF", colour = NA))


  ggsave(CHX_effect_93, file = file.path(results_path, paste0("final_sample_GSE_OnlyTreatmenteffect.tif")), dpi = 600, width = 13, height = 6)
  
  CHX_effect_93
}


```
